Fix `fatbuildr.tasks.manager error while running task <id>: maximum recursion depth exceeded`

Fix this error as reported in `fatbuildrd` logs:

```
fatbuildrd[406734]: worker-default: [INFO] fatbuildr.builds Running build 2eb06164-09d9-47dc-8139-17d86327fb53
fatbuildrd[406734]: worker-default: [INFO] fatbuildr.builds Extracting tarball /run/fatbuildr/e2zjmq90/fatbuildr-artifact.tar.xz in destination /var/lib/fatbuildr/workspaces/default/2eb06164-09d9-47dc-8139-17d86327>
fatbuildrd[406734]: worker-default: [DEBUG] fatbuildr.builds Removing build input tarball /run/fatbuildr/e2zjmq90/fatbuildr-artifact.tar.xz
fatbuildrd[406734]: worker-default: [DEBUG] fatbuildr.artifact Loading artifact definitions from /var/lib/fatbuildr/workspaces/default/2eb06164-09d9-47dc-8139-17d86327fb53/artifact.yml
fatbuildrd[406734]: worker-default: [INFO] fatbuildr.builds Using provided source archive /var/lib/fatbuildr/workspaces/default/2eb06164-09d9-47dc-8139-17d86327fb53/firehpc-1.2.0a1.dev49.tar.xz
fatbuildrd[406734]: worker-default: [DEBUG] fatbuildr.builds Artifact main version extracted from source tarball name:  1.2.0a1.dev49
fatbuildrd[406734]: worker-default: [DEBUG] fatbuildr.builds Removing build submission temporary files directory /run/fatbuildr/e2zjmq90
fatbuildrd[406734]: worker-default: [INFO] fatbuildr.builds Using artifact source archive from cache /var/cache/fatbuildr/default/firehpc/v0.0.1.tar.gz
fatbuildrd[406734]: worker-default: [DEBUG] fatbuildr.utils Extracting main source archive /var/lib/fatbuildr/workspaces/default/2eb06164-09d9-47dc-8139-17d86327fb53/firehpc-1.2.0a1.dev49.tar.xz in /var/lib/fatbuil>
fatbuildrd[406734]: worker-default: [DEBUG] fatbuildr.utils Extracting supplementary source archive /var/cache/fatbuildr/default/firehpc/v0.0.1.tar.gz in /var/lib/fatbuildr/workspaces/default/2eb06164-09d9-47dc-813>
fatbuildrd[406734]: worker-default: [INFO] fatbuildr.archive skipping extraction of file ansible-connection-machinectl-0.0.1 due to stripping
fatbuildrd[406734]: worker-default: [ERROR] fatbuildr.tasks.manager error while running task 2eb06164-09d9-47dc-8139-17d86327fb53: maximum recursion depth exceeded
fatbuildrd[406734]: worker-default: [INFO] fatbuildr.tasks.manager Task failed
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -8,6 +8,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ## [Unreleased]
 
 ### Fixed
+- Fix infinite recursion error with `PatchesSubdir` on Python 3.12+ (#195).
 - docs: Add missing `sudo` matrix hooks installation procedure.
 
 ## [2.1.0] - 2024-07-16

--- a/fatbuildr/builds/__init__.py
+++ b/fatbuildr/builds/__init__.py
@@ -365,14 +365,7 @@ class ArtifactBuild(RunnableTask):
         # Render all patches templates found in patches directories
         for patch in self.patches:
             if patch.template:
-                patch_tmp = patch.with_suffix('.swp')
-                patch.rename(patch_tmp)
-                logger.info("Rendering patch template %s", patch)
-                with open(patch, 'w+') as fh:
-                    fh.write(
-                        Templeter().frender(patch_tmp, version=self.version)
-                    )
-                patch_tmp.unlink()
+                patch.render(version=self.version)
 
         # Render rename index template if present
         rename_idx_path = self.place.joinpath('rename')

--- a/fatbuildr/git.py
+++ b/fatbuildr/git.py
@@ -38,6 +38,7 @@ except ImportError:
 import pygit2
 from debian import deb822
 
+from .templates import Templeter
 from .log import logr
 
 logger = logr(__name__)
@@ -56,32 +57,23 @@ def is_meta_generic(meta):
 def load_git_repository(path: str):
     return pygit2.Repository(path)
 
-
-# It is not possible to inherit directory from pathlib.Path as this class
-# constructor actually returns the specialized type according to the underlying
-# OS. One solution is to discover the actual specialized type for the current OS
-# and inherit from it.
-PATH_TYPE = type(Path())
-
-
-class PatchesDir(PATH_TYPE):
-    """Class to manipulatz patches directories and their subdirs. In pathlib
+class PatchesDir:
+    """Class to manipulate patches directories and their subdirs. In pathlib
     classes the attributes are given as arguments to the constructor, and not to
     the __init__() method. This class overrides the constructor to add the
     version argument."""
 
-    def __new__(cls, path, version):
-        obj = super().__new__(cls, path.joinpath('patches'))
-        obj.version = version
-        return obj
+    def __init__(self, path, version):
+        self._path = Path(path.joinpath('patches'))
+        self.version = version
 
     @cached_property
     def generic_subdir(self):
-        return PatchesSubdir(self, self.joinpath('generic'))
+        return PatchesSubdir(self, 'generic')
 
     @cached_property
     def version_subdir(self):
-        return PatchesSubdir(self, self.joinpath(self.version))
+        return PatchesSubdir(self, self.version)
 
     @property
     def subdirs(self):
@@ -97,66 +89,78 @@ class PatchesDir(PATH_TYPE):
 
     def ensure(self):
         """Create patches directory if it does not exist yet."""
-        if not self.exists():
-            logger.debug("Creating artifact patches directory %s", self)
-            self.mkdir()
-            self.chmod(0o755)
+        if not self._path.exists():
+            logger.debug("Creating artifact patches directory %s", self._path)
+            self._path.mkdir()
+            self._path.chmod(0o755)
 
 
-class PatchesSubdir(PATH_TYPE):
+class PatchesSubdir:
     """Class to manipulate patches subdirectories. In pathlib classes the
     attributes are given as arguments to the constructor, and not to the
     __init__() method. This class overrides the constructor to add the
     patches_dir argument."""
 
-    def __new__(cls, patches_dir, path):
-        obj = super().__new__(cls, path)
-        obj.patches_dir = patches_dir
-        return obj
+    def __init__(self, patches_dir, sub_dir):
+        self.patches_dir = patches_dir
+        self._path = self.patches_dir._path.joinpath(sub_dir)
 
     @property
     def patches(self):
         """Return sorted list of PatchFiles available in subdir."""
-        return sorted([PatchFile(patch) for patch in self.iterdir()])
+        return sorted([PatchFile(patch) for patch in self._path.iterdir()])
+
+    def exists(self):
+        return self._path.exists()
 
     def ensure(self):
         """Create patches subdirectory if it does not exist yet."""
         # First ensure parent patches directory
         self.patches_dir.ensure()
 
-        if not self.exists():
-            logger.debug("Creating patches subdirectory %s", self)
-            self.mkdir()
-            self.chmod(0o755)
+        if not self._path.exists():
+            logger.debug("Creating patches subdirectory %s", self._path)
+            self._path.mkdir()
+            self._path.chmod(0o755)
 
     def clean(self):
         """Remove all existing patches in subdir."""
-        if not self.exists():
+        if not self._path.exists():
             return
         for patch in self.patches:
             logger.debug("Removing old patch %s", patch.fullname)
-            patch.unlink()
+            patch.remove()
 
 
-class PatchFile(PATH_TYPE):
+class PatchFile:
     """Class to manipulate patch files."""
 
     TEMPLATE_KEY = 'Template'
 
+    def __init__(self, path: Path):
+        self._path = path
+
+    def __lt__(self, other):
+        return self._path < other._path
+
+    @property
+    def name(self):
+        return self._path.name
+
     @property
     def fullname(self):
-        return f"{self.parent.name}/{self.name}"
+        return f"{self._path.parent.name}/{self._path.name}"
 
     @cached_property
     def content(self):
-        with open(self, 'rb') as fh:
+        with open(self._path, 'rb') as fh:
             return fh.read()
 
     @cached_property
     def title(self):
         # Extract commit title (1st line) from patch filename, by removing
         # the patch index before the first dash.
-        return self.name.split('-', 1)[1]
+        return self._path.name.split('-', 1)[1]
 
     @cached_property
     def meta(self):
@@ -171,6 +175,16 @@ class PatchFile(PATH_TYPE):
             and self.meta[self.TEMPLATE_KEY] == 'yes'
         )
 
+    def render(self, **kwargs):
+        patch_tmp = self._path.with_suffix('.swp')
+        self._path.rename(patch_tmp)
+        logger.info("Rendering patch template %s", self._path)
+        with open(self._path, 'w+') as fh:
+            fh.write(
+                Templeter().frender(patch_tmp, **kwargs)
+            )
+        patch_tmp.unlink()
+
     def in_field(self, field, value):
         """Returns True if value in found in space separated list field, or
         False otherwise."""
@@ -178,16 +192,22 @@ class PatchFile(PATH_TYPE):
 
     @property
     def generic(self):
-        return self.parent.name == 'generic'
+        return self._path.parent.name == 'generic'
 
     @staticmethod
     def create(subdir, title):
-        return PatchFile(subdir.joinpath(title))
+        return PatchFile(subdir._path.joinpath(title))
 
     def write(self, meta, diff):
-        with open(self, 'w+') as fh:
+        with open(self._path, 'w+') as fh:
             fh.write(str(meta) + '\n\n')
             fh.write(diff)
+
+    def rename(self, to):
+        self._path.rename(to)
+
+    def remove(self):
+        self._path.unlink()
 
 
 class GitRepository:

--- a/tests/test_git.py
+++ b/tests/test_git.py
@@ -21,6 +21,9 @@ import unittest
 import tempfile
 from pathlib import Path
 import shutil
+import textwrap
+
+import deb822
 
 from fatbuildr.git import PatchesDir, PatchesSubdir, PatchFile
 
@@ -84,3 +87,143 @@ class TestPatchesSubdir(unittest.TestCase):
         self.assertFalse(self.patches_subdir.exists())
         self.patches_subdir.ensure()
         self.assertTrue(self.patches_subdir.exists())
+
+class TestPatchFile(unittest.TestCase):
+    def setUp(self):
+        self.test_dir = tempfile.mkdtemp()
+        self.patchesdir = PatchesDir(Path(self.test_dir), "1.0.0")
+        (self.generic_subdir, self.version_subdir) = self.patchesdir.subdirs
+        # Version specific patch
+        self.path = self.version_subdir._path.joinpath("0000-test.patch")
+        self.content = textwrap.dedent(
+            """
+            Description: Very very long multi-line
+             description.
+            Author: John Doe <john.doe@corp.org>
+            Last-Update: 1970-01-02
+            """
+        )
+        self.diff = textwrap.dedent(
+            """
+            diff --git a/source.py b/source.py
+            index 3b1399e..84699ca 100644
+            --- a/source.py
+            +++ b/source.py
+            @@ -1 +1 @@
+            -BUG
+            +FIX
+            """
+        )
+        self.version_subdir.ensure()
+        with open(self.path, "w+") as fh:
+            fh.write(self.content + "\n\n" + self.diff)
+        self.patch = PatchFile(self.path)
+
+        # Generic subdir
+        self.template_path = self.generic_subdir._path.joinpath(
+            "0001-template.patch"
+        )
+        self.template_content = textwrap.dedent(
+            """
+            Description: Template path.
+            Author: John Doe <john.doe@corp.org>
+            Last-Update: 1970-01-03
+            Template: yes
+            """
+        )
+        self.template_diff = textwrap.dedent(
+            """
+            diff --git a/source.py b/source.py
+            index 3b1399e..84699ca 100644
+            --- a/source.py
+            +++ b/source.py
+            @@ -1 +1 @@
+            -BUG
+            +{{ version }}
+            """
+        )
+        self.generic_subdir.ensure()
+        with open(self.template_path, "w+") as fh:
+            fh.write(self.template_content + "\n\n" + self.template_diff)
+        self.template_patch = PatchFile(self.template_path)
+
+    def tearDown(self):
+        shutil.rmtree(self.test_dir)
+
+    def test_name(self):
+        self.assertEqual(self.patch.name, "0000-test.patch")
+        self.assertEqual(self.template_patch.name, "0001-template.patch")
+
+    def test_fullname(self):
+        self.assertEqual(self.patch.fullname, "1.0.0/0000-test.patch")
+        self.assertEqual(
+            self.template_patch.fullname, "generic/0001-template.patch"
+        )
+
+    def test_content(self):
+        self.assertEqual(
+            self.patch.content, (self.content + "\n\n" + self.diff).encode()
+        )
+
+    def test_title(self):
+        self.assertEqual(self.patch.title, "test.patch")
+        self.assertEqual(self.template_patch.title, "template.patch")
+
+    def test_meta(self):
+        self.assertEqual(
+            self.patch.meta["Description"],
+            "Very very long multi-line\n description."
+        )
+        self.assertEqual(
+            self.patch.meta["Author"], "John Doe <john.doe@corp.org>"
+        )
+        self.assertEqual(
+            self.patch.meta["Last-Update"], "1970-01-02"
+        )
+        # Test undefined key
+        self.assertNotIn("Unknown", self.patch.meta)
+        with self.assertRaises(KeyError):
+            self.patch.meta["Unknown"]
+
+    def test_template(self):
+        self.assertFalse(self.patch.template)
+        self.assertTrue(self.template_patch.template)
+
+    def render(self):
+        self.template_patch.render(version="3.2.1")
+        self.assertTrue(
+            self.template_patch.content().decode().endswith("3.2.1\n")
+        )
+
+    def test_in_field(self):
+        self.assertTrue(self.patch.in_field("Author", "John"))
+        self.assertFalse(self.patch.in_field("Description", "short"))
+        with self.assertRaises(KeyError):
+            self.patch.in_field("Unknown", "whatever")
+
+    def test_generic(self):
+        self.assertFalse(self.patch.generic)
+        self.assertTrue(self.template_patch.generic)
+
+    def test_create(self):
+        patch = PatchFile.create(self.generic_subdir, "0002-creation-test")
+        self.assertIsInstance(patch, PatchFile)
+        self.assertEqual(patch.title, "creation-test")
+
+    def test_write(self):
+        meta = deb822.Deb822()
+        meta["Author"] = "Jane Doe <jane.doe@corp.org>"
+        self.patch.write(meta, self.diff)
+        self.assertEqual(
+            self.patch.meta["Author"],
+            "Jane Doe <jane.doe@corp.org>"
+        )
+
+    def test_rename(self):
+        to = self.generic_subdir._path.joinpath("0002-renamed.patch")
+        self.patch.rename(to)
+        self.assertTrue(to.exists())
+
+    def test_remove(self):
+        self.patch.remove()
+        self.assertFalse(self.path.exists())

--- a/tests/test_git.py
+++ b/tests/test_git.py
@@ -1,0 +1,62 @@
+#!/usr/bin/env python3
+#
+# Copyright (C) 2024 Rackslab
+#
+# This file is part of Fatbuildr.
+#
+# Fatbuildr is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# Fatbuildr is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Fatbuildr.  If not, see <https://www.gnu.org/licenses/>.
+
+import unittest
+import tempfile
+from pathlib import Path
+import shutil
+
+from fatbuildr.git import PatchesDir
+
+class TestPatchesDir(unittest.TestCase):
+    def setUp(self):
+        self.test_dir = tempfile.mkdtemp()
+        self.patchesdir = PatchesDir(Path(self.test_dir), "1.0.0")
+
+    def tearDown(self):
+        shutil.rmtree(self.test_dir)
+
+    def test_empty(self):
+        # Check initially empty
+        self.assertTrue(self.patchesdir.empty)
+
+        # Create one subdir and an empty patch
+        subdir = self.patchesdir.version_subdir
+        subdir.ensure()
+        with open(subdir._path.joinpath("test.path"), "w+"):
+            pass
+
+        # Check empty is now false
+        self.assertFalse(self.patchesdir.empty)
+
+    def test_path(self):
+        self.assertEqual(str(self.patchesdir._path), f"{self.test_dir}/patches")
+
+    def test_version_subdir(self):
+        subdir = self.patchesdir.version_subdir
+        self.assertEqual(str(subdir._path), f"{self.test_dir}/patches/1.0.0")
+
+    def test_generic_subdir(self):
+        subdir = self.patchesdir.generic_subdir
+        self.assertEqual(str(subdir._path), f"{self.test_dir}/patches/generic")
+
+    def test_subdirs(self):
+        (generic_subdir, version_subdir) = self.patchesdir.subdirs
+        self.assertEqual(generic_subdir, self.patchesdir.generic_subdir)
+        self.assertEqual(version_subdir, self.patchesdir.version_subdir)

--- a/tests/test_git.py
+++ b/tests/test_git.py
@@ -22,7 +22,7 @@ import tempfile
 from pathlib import Path
 import shutil
 
-from fatbuildr.git import PatchesDir
+from fatbuildr.git import PatchesDir, PatchesSubdir, PatchFile
 
 class TestPatchesDir(unittest.TestCase):
     def setUp(self):
@@ -60,3 +60,27 @@ class TestPatchesDir(unittest.TestCase):
         (generic_subdir, version_subdir) = self.patchesdir.subdirs
         self.assertEqual(generic_subdir, self.patchesdir.generic_subdir)
         self.assertEqual(version_subdir, self.patchesdir.version_subdir)
+
+class TestPatchesSubdir(unittest.TestCase):
+    def setUp(self):
+        self.test_dir = tempfile.mkdtemp()
+        self.patches_dir = PatchesDir(Path(self.test_dir), "1.0.0")
+        self.patches_subdir = PatchesSubdir(self.patches_dir, "generic")
+
+    def tearDown(self):
+        shutil.rmtree(self.test_dir)
+
+    def test_patches(self):
+        self.patches_subdir.ensure()
+        self.assertCountEqual(self.patches_subdir.patches, [])
+        with open(self.patches_subdir._path.joinpath("test.patch"), "w+"):
+            pass
+        self.assertEqual(len(self.patches_subdir.patches), 1)
+        self.assertIsInstance(self.patches_subdir.patches[0], PatchFile)
+        self.patches_subdir.clean()
+        self.assertCountEqual(self.patches_subdir.patches, [])
+
+    def test_exists(self):
+        self.assertFalse(self.patches_subdir.exists())
+        self.patches_subdir.ensure()
+        self.assertTrue(self.patches_subdir.exists())

